File: java/hash_simples/Hash.java
package hash_simples;

import java.util.List;

public class Hash {
    public  int tableSize;
    public  int overflowSize;
    private int realsize;
    private int posOverflow; // control insertions on overflow area
    public  Object htable[]; // make easy to substitute the type of the table

    public Hash() {
        this(10, 10);
    }

    public Hash(int tableSize, int overflowSize) {
		// if the size is invalid, both will 
		// have the value of 10
		if (isSizeValid(tableSize))
			this.tableSize = tableSize;
		else {
			System.out.println("Tamanho de área principal inválido. Assumindo o valor 10.");
			this.tableSize = 10;
		}

		if (isSizeValid(overflowSize))
			this.overflowSize = overflowSize;
		else {
			System.out.println("Tamanho de área reserva inválido. Assumindo o valor 10.");
			this.overflowSize = 10;
		}

        realsize = tableSize + overflowSize;
        posOverflow = 0;
        htable = new Object[realsize];

        for (int i = 0; i < realsize; i++) 
            htable[i] = null;
    }

	private boolean isSizeValid(int size) {
		return (size > 0);
	}

    private int hash(Object item) {
        // change the key with whatever you want
        return ((int)item % tableSize);
    }

    // return the position where item has been inserted
    public int insert(Object item) {
        int pInsertion = hash(item); 
        int posReturn = -1; 
        // if the position is free
        if (htable[pInsertion] == null) {
            htable[pInsertion] = item;
            posReturn = pInsertion;
        } else {
            if (posOverflow < (realsize - tableSize)) {
                htable[tableSize + posOverflow] = item;
                posReturn = tableSize + posOverflow;
                posOverflow++;
            } else {
                System.out.println("Área de reserva cheia! Falha ao inserir "+(int)item);
            }
        }
        return posReturn;
    }

	public void insertList(List<Integer> lista) {
		for (int item : lista) 
			insert((Object) item);
	}

    // return the position of the removed element
    // -1 if not found
    public int remove(Object toBeRemoved) {
        int pos = -1;
        int pRemotion = hash(toBeRemoved);
        // found in the main area
        if (htable[pRemotion] != null && ((int)htable[pRemotion] == (int)toBeRemoved)) {
            pos = pRemotion;
            boolean hasSubstitute = false;
            // change the item in the main area
            // with the item in the overflow area
            // with the same hash 
            for (int i = 0; i < posOverflow && !hasSubstitute; i++) {
                if (hash(htable[i + tableSize]) == pRemotion) {
                    hasSubstitute = true;
                    htable[pRemotion] = htable[i + tableSize];
                    if (posOverflow > 0) 
                        posOverflow--;
                    htable[i + tableSize] = htable[tableSize + posOverflow];
                    htable[tableSize + posOverflow] = null;
                }
            }
			if (!hasSubstitute)
				htable[pRemotion] = null;
        } else {
            // not found in main area -> search in the overflow area
            for (int i = 0; i < posOverflow; i++) {
                if ((int)htable[i + tableSize] == (int)toBeRemoved) {
                    pos = i + tableSize;
                    if (posOverflow > 0)
                        posOverflow--;
                    htable[i + tableSize] = htable[posOverflow];
                    htable[posOverflow] = null;
                    i = posOverflow;
                }
            }
        }
        return pos;
    }

    // return the removed item
    // null if not found or if the key is invalid
    public Object remove(int key) {
        if (key < 0 || key >= realsize) {
            System.out.println("Chave de remoção inválida.");
            return null;
        }
        Object removed = null;
		// if there is something in the position
        if (htable[key] != null) {
            removed = htable[key];
            boolean hasSubstitute = false;
            // again, put in the main area
            // the item with the same hash coming
            // from the overflow area
            for (int i = 0; i < posOverflow && !hasSubstitute; i++) {
                if (hash(htable[i + tableSize]) == key) {
                    hasSubstitute = true;
                    htable[key] = htable[i + tableSize];
                    if (posOverflow > 0) 
                        posOverflow--;
                    htable[i + tableSize] = htable[posOverflow];
                    htable[posOverflow] = null;
                }
            }
			if (!hasSubstitute)
				htable[key] = null;
        }
        return removed;
    }

    public void showItens() {
        System.out.println("\nÁrea principal: [");
        for (int i = 0; i < tableSize; i++) {
            if (htable[i] != null)
                System.out.println("    Índice "+i+":  "+(int)htable[i]);
            else
                System.out.println("    Índice "+i+":  vazio");
        }
        System.out.println("]\n");
        System.out.println("Área de reserva: [");
        for (int i = 0; i < (realsize - overflowSize); i++) {
            if (htable[i + tableSize] != null)
				System.out.print("    Índice "+(i+tableSize)+":  "+(int)htable[tableSize + i] + "\n");
            else
                System.out.println("    Índice "+(i+tableSize)+":  vazio");
        }
        System.out.println("]\n");
    }

    public boolean search(Object obj) {
        boolean retorno = false;
        if (obj == null) {
            System.out.println("Objeto inválido.");
        } else {
            int hashJ = hash(obj);
            if (htable[hashJ] != null && ((int)htable[hashJ] == (int)obj)) {
                retorno = true;
            } else {
                // searching on overflow 
                for (int i = 0; i < posOverflow && !retorno; i++) {
                    if (htable[i + tableSize] != null && ((int)htable[i + tableSize] == (int)obj)) {
                        retorno = true;
					}
                }
            }
        }
        return retorno;
    }

    // return the position of the item
    // -1 if not found
    public int searchPos(Object obj) {
        int pos = -1;
        if (obj == null) {
            System.out.println("Objeto inválido.");
        } else {
            int hashJ = hash(obj);
            if (htable[hashJ] != null && ((int)htable[hashJ] == (int)obj)) {
                pos = hashJ;
            } else {
                // searching in overflow area
                for (int i = 0; i < posOverflow && (pos == -1); i++) {
                    if (htable[i + tableSize] != null && ((int)htable[i + tableSize] == (int)obj)) {
                        pos = i + tableSize;
                    }
                }
            }
        }
        return pos;
    }
}

File: c/hash_simples/hash.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

#include "hash.h"
#include "menuHashSimples.h"

Hash * newHash(int tamTable, int tamOverflow) {
	Hash * th = (Hash *) malloc(sizeof(Hash));

	th->posOverflow = 0;
	th->tableSize = tamTable;
	th->overflowSize = tamOverflow;
	th->realSize = tamTable + tamOverflow;

    th->tabela = (int **) malloc(sizeof(int *) * th->realSize);

    for (int i = 0; i < th->realSize; i++) {
        th->tabela[i] = NULL;
    }

    return th;
}

int hash(Hash * th, int key) {
    return key % th->tableSize;
}

// return the position where the item
// has been inserted
// return -1 if could not insert
int inserir(Hash * th, int item) {
    int posInsercao = -1;
    int pos = hash(th, item);
    int *p = (int *) malloc(sizeof(int));
    *p = item;
    if (th->tabela[pos] == NULL) {
        th->tabela[pos] = p;
        posInsercao = pos;
    } else {
		// check if there is space to insert something
		// in the overflow area
        if (th->posOverflow < (th->realSize - th->tableSize)) {
            int posReserva = th->tableSize + th->posOverflow++;
            th->tabela[posReserva] = p; 
            posInsercao = posReserva;
        } else {
            printf("Area de reserva cheia!\n");
            free(p);
        }
    }
    p = NULL;
    return posInsercao;
}

// remove item using its position
int removerPos(Hash * th, int key) {
    int removed = 0x7fffffff;

    // is key valid?
    if (key < 0 || key >= th->realSize) {
        printf("Chave de procura invalida. Chave: %d\n", key);
    } else {
        // if there is something in the position
        if (th->tabela[key] != NULL) {
            removed = *th->tabela[key];
            bool hasSubstitute = false;
			// search for a substitute that has
			// the same hash of the item
            for (int i = 0; i < th->posOverflow && !hasSubstitute; i++) {
                if (hash(th, *th->tabela[i + th->tableSize]) == key) {
                    hasSubstitute = true;
                    *(th->tabela[key]) = *(th->tabela[i + th->tableSize]);
                    if (th->posOverflow > 0)
                        th->posOverflow--;
                    *(th->tabela[i + th->tableSize]) = *(th->tabela[th->tableSize + th->posOverflow]);
                    free(th->tabela[th->tableSize + th->posOverflow]);
                    th->tabela[th->tableSize + th->posOverflow] = NULL;
                }
            }
			// if there is no substitute,
			// just remove the item
            if (!hasSubstitute) {
                free(th->tabela[key]);
                th->tabela[key] = NULL;
            }
        }
    }

    return removed;
}

// remove item -> return its position 
int removerItem(Hash * th, int item) {
    int posRemovido = -1;
    int posItem = hash(th, item);
    if (*(th->tabela[posItem]) == item) {
        posRemovido = posItem;
        bool hasSubstitute = false;
        for (int i = 0; i < th->posOverflow && !hasSubstitute; i++) {
            if (hash(th, *(th->tabela[i + th->tableSize])) == posItem) {
                hasSubstitute = true;
                *(th->tabela[posItem]) = *(th->tabela[i + th->tableSize]);
                if (th->posOverflow > 0)
                    th->posOverflow--;
                *(th->tabela[i + th->tableSize]) = *(th->tabela[th->tableSize + th->posOverflow]);
                free(th->tabela[th->tableSize + th->posOverflow]);
                th->tabela[th->tableSize + th->posOverflow] = NULL;
            }
        }
        if (!hasSubstitute) {
            free(th->tabela[posItem]);
            th->tabela[posItem] = NULL;
        }
    } else {
        // searching in the overflow area
        for (int i = 0; i < th->posOverflow; i++) {
            if (*(th->tabela[i + th->tableSize]) == item) {
                posRemovido = i + th->tableSize;
                // if th->posOverflow > 0, the position of
                // the last item in overflow area
                // is th->tableSize + th->posOverflow - 1
                if (th->posOverflow > 0)
                    th->posOverflow--;
                *(th->tabela[i + th->tableSize]) = *(th->tabela[th->tableSize + th->posOverflow]);
                free(th->tabela[th->tableSize + th->posOverflow]);
                th->tabela[th->tableSize + th->posOverflow] = NULL;
                i = th->posOverflow;
            }
        }
    }
    return posRemovido;
}

void mostrar(Hash * th) {
    printf("\nÁrea principal: [ ");
    for (int i = 0; i < th->tableSize; i++) { 
		if (th->tabela[i] != NULL) 
			printf("%d ", *(th->tabela[i]));
	}
    printf("]\n");
    
    printf("Área de reserva: [ ");
    for (int i = 0; i < th->overflowSize; i++) {
		if ((th->tabela[ i + (th->tableSize) ]) != NULL)
			printf("%d ", *(th->tabela[i + th->tableSize]));
	}
    printf("]\n\n");
}

void desalocarTabela(Hash * th) {
	for (int i = 0; i < th->realSize; i++) {
		free(th->tabela[i]);
	}
	free(th);
}
